resource/resourceadapters/charmstore_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resourceadapters_test
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/charmstore"
+	"github.com/juju/juju/resource/resourceadapters"
+)
+
+type CharmStoreSuite struct {
+	testing.IsolationSuite
+
+	resourceClient *testResourceClient
+}
+
+var _ = gc.Suite(&CharmStoreSuite{})
+
+func (s *CharmStoreSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+	s.resourceClient = &testResourceClient{
+		stub: &testing.Stub{},
+	}
+}
+
+func (s *CharmStoreSuite) TestGetResourceTerminates(c *gc.C) {
+	msg := "trust"
+	attempts := 0
+	s.resourceClient.getResourceF = func(req charmstore.ResourceRequest) (data charmstore.ResourceData, err error) {
+		attempts++
+		return charmstore.ResourceData{}, errors.New(msg)
+	}
+	csRes := resourceadapters.NewCSRetryClientForTest(s.resourceClient)
+
+	_, err := csRes.GetResource(charmstore.ResourceRequest{})
+	c.Assert(err, gc.ErrorMatches, fmt.Sprintf("failed after retrying: %v", msg))
+	// Ensure we logged attempts @ WARNING.
+	c.Assert(c.GetTestLog(), jc.Contains, fmt.Sprintf("WARNING juju.resource.resourceadapters (attempt %v) retrying resource ", attempts))
+
+	callsMade := []string{}
+	for i := 0; i < attempts; i++ {
+		callsMade = append(callsMade, "GetResource")
+	}
+	s.resourceClient.stub.CheckCallNames(c, callsMade...)
```

Comment:
I was not checking since the number of GetResources should match attempts and am checking how many times we call GetResources. I have added the check though